editors/code/src/ctx.ts

```diff
@@ -26,16 +29,14 @@ export class Ctx {
 
     get activeRustEditor(): RustEditor | undefined {
         const editor = vscode.window.activeTextEditor;
-        return editor && isRustEditor(editor)
-            ? editor
-            : undefined;
+        return editor && isRustEditor(editor) ? editor : undefined;
     }
 
     get visibleRustEditors(): RustEditor[] {
         return vscode.window.visibleTextEditors.filter(isRustEditor);
     }
 
-    registerCommand(name: string, factory: (ctx: Ctx) => Cmd) {
+    registerCommand<T extends unknown[]>(name: string, factory: (ctx: Ctx) => Cmd<T>): void {
```

Comment:
Strongly-typing the `Cmd` doesn't give us many benefits but gives only increased verbosity and repetition (the passed parameters tuple type just unnecessarily duplicates the type of lambda parameters in command handlers). VSCode doesn't provide the strongly-typed interface for commands so we won't get compile errors for them. This is actually one of the rare cases where we should use `any` as the function parameters type.

Reply:
And if I could, I'd remove this interface from return types and rely on type inference, so we don't cast strongly-typed lambdas to type-erased `Cmd`, e.g.:
```ts
function createCmdHandler(): Cmd {
    return (a: number, b: string) => { /**/ }
}
createCmdHandler()(null, null); // gives no error
```
vs when infered
```ts
function createCmdHandler() {
    return (a: number, b: string) => { /* */ }
}
createCmdHandler()(null, null); // compile error
```

Reply:
Yeah, I don't disagree with your thoughts about the strongly typed `Cmd`. In fact, I agree it doesn't really add much in practice and I almost didn't include this change.

I just really kind of dislike using `any`. It often seems harmless but can lead to surprising and unfortunate bugs from time to time. I suppose if it's just for the definition of `Cmd` and it only needs the one use of `// eslint-disable @typescript-eslint/no-explicit-any`, it's not so bad.

One option which might be acceptable would be to just remove the type annotations from the lambda arguments and only specify them in `T` for `Cmd<T>`. That way we don't need any, we don't lose any type information (argument types are still inferrable), and it's in the end not much more verbose.

Example:

```typescript
export function showReferences(ctx: Ctx): Cmd<[string, lc.Position, lc.Location[]]> {
    return (uri: string, position: lc.Position, locations: lc.Location[]): void => {
```

would look like

```typescript
export function showReferences(ctx: Ctx): Cmd<[string, lc.Position, lc.Location[]]> {
    return (uri, position, locations): void => {
```

What do you think? If you prefer, I can still switch back to `any`.

Reply:
Yeah, I am glad, I found one more like-minded person who is passionate about static typing. I do think that `any` always leads to bugs. I wish all the typescript code avoided it.

However, it is unavoidable in your code due to someone's bad design decisions (e.g. https://github.com/rust-analyzer/rust-analyzer/pull/3629#discussion_r394311242, and untyped vscode commands api just forces you to deal with `any`) and bad language design (e.g. `catch` error variable is always of `any` type and there is even no way to redefine it to `unknown`, but I am getting too far...).

Though I do like relying on return type inference, not everyone likes it. So, in this case, your proposal does seem viable, so we do get rid of the dreaded `any` at least in our code.